### PR TITLE
Allow removing shares which are invalid due to missing source fileid

### PR DESCRIPTION
**Pull request by jvillafanez**
## Related Issue
Related to https://github.com/owncloud/windows_network_drive/pull/310

## Motivation and Context
clean up data that won't be accessible

## How Has This Been Tested?
Checked along with https://github.com/owncloud/windows_network_drive/pull/310

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by jvillafanez** [comment]
Code-wise this is ready, but we might need to adjust the "since" tags. I'm assuming this will be for 10.7

**Comment by mmattel** [comment]
@jvillafanez any progress on that?
Could you add a changelog pls?

**Comment by jvillafanez** [comment]
Since this PR has API changes, I'm waiting for the official 10.7 alpha. Master is still with 10.6

File: changelog/unreleased/38152 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Change: API changes to remove shares pointing to missing files
+
+If a file was completely deleted without unsharing first, the share
+would still exist in the DB even though it wouldn't be shown to the users.
+This change opens a way to remove those shares.
```
Comment (mmattel) [inline changelog/unreleased/38152:5]: The text reads like a user/admin has the capability to do that. As I can´t see a occ command for doing so, this seems not to be the case. Therefore I propose to rewrite the text a bit to make it more clear how this is initiated.

Comment (jvillafanez) [reply changelog/unreleased/38152:5]: For now, the windows_network_drive app is the only app that will use this feature (in a future release), so I'd rather not mention it here.
Other than that, this feature isn't used at the moment. Including an occ command could be useful, but it needs to be discussed.

Comment (mmattel) [reply changelog/unreleased/38152:5]: Maybe: ‘This change opens‘ —> ‘This change prepares‘

**Comment by jvillafanez** [comment]
This is currently frozen due to API changes. Waiting for OC 10.7 in order to resume the work in this area. There is a 10.6.1 version planned I think.
@micbar FYI

**Comment by phil-davis** [comment]
> There is a 10.6.1 version planned I think.

Note: there is no 10.6.1 planned. Next version should be 10.7.0

**Comment by VicDeo** [comment]
@jvillafanez looks good in general. 
Is it possible not to duplicate the code for the core federatedfilesharing app? As I see the only difference here is a set of share types.

**Comment by jvillafanez** [comment]
> Is it possible not to duplicate the code for the core federatedfilesharing app? As I see the only difference here is a set of share types.

The problem is that the actual action depends on each provider. The manager just forwards the request to all the providers. In this context, we could expect that each provider has its own way of storing the information, and its own way to delete it. The fact that both providers use the same DB table could be considered as a bad coincidence.

I agree we should avoid that duplication, but I think we'd also need to refactor all the related operations and provide a common way for the providers to store the information. For now, as said, it seems the storing part is completely delegated to the providers.

**Review by VicDeo — APPROVED** [review]

**Review by AlexAndBear — APPROVED** [review]

**Pull request merged**